### ISSUE TITLE
Promisified invocation of listeners

### ISSUE THREAD

**Issue by fabien** [router5/router5#3]
I'm wondering, wouldn't it be highly desirable to (optionally) use promises to handle any async operations (animated transitions, server calls) within event handlers? This would fit in nicely with the plan to support async `canDeactivate` handling.

**Comment by troch** [comment]
Thanks for taking interest in this project :smile: 

The way I see router5 being used: on a route change, an action is passed to a store which loads data and the view is at the receiving end (flux-like implementation). It is not designed to control async data flow, and I prefer to allow maximum flexibility.

For listeners, I have not limited the number of listeners which can be registered (either on a node, on a specific route, or on any route change), and they are not called in a specific order. To me the use of promises would only be used by the router to apply / not apply a route change (the result of a transition). Control flow (of async data loading) is up to the user rather than the router.

A solution to support this would be to give the possibility to register a `canActivate` method per segment. Like `canDeactivate`, it could support promises and would be pipelined from top to bottom (bottom to top for _canDeactivate_).

What would be your use case for the router to use promises other than to seek activation / deactivation authorisation?

**Comment by fabien** [comment]
Thanks for your prompt reply! My main concern would be that the user is not blocked from navigating while async operations would still be pending/running. The url would probably already have changed, while the UI hasn't caught up with the representation of this state.

Note that I'm actually planning to use router5 in a Backbone/Marionette setup, so maybe I misunderstood what your React/Deku examples do right now. 

I think it would be helpful to have a more bare-bones (perhaps even plain JS/DOM) example of how the router would be used. Ideally, the example(s) would incorporate async flow, like UI transitions or data loading.

I was almost tempted to wrap your invokeListeners calls in promises/deferred logic today, to see if I could make that work after all. But apparently they are not called in a particular order, so I guess that's a no-go?

Is there a chance to have your `canActivate` support promises anytime soon?

**Comment by troch** [comment]
__Pipelined promises__

`canDeactivate` will support promises soon, and I'll add `canActivate` at the same time (I plan to do it this week).

__URL change__

I myself don't mind having an URL change early: this is how browsers work (you enter your URL first and get the page later), and it allows a refresh during the transition phase without going back to the previous state / view (like if an action of pressing a button never happened).

However I don't want to push people in a particular way as there are different views on this. To me the most important listener is the one you register with `addNodeListener`, because it is the one which will trigger a view update. I should limit node listeners to 1 per node. That way I could invoke them first and handle a promise they may return before invoking other listeners.

A route change could then be (if any step is unsuccessful, transition is aborted):
- Invoke `canDeactivate` from bottom to top for deactivating segments
- Invoke `canActivate` from top to bottom for activating segments
- Invoke listener registered on the intersection node of deactivating / activating segments
- Update URL immediately if no promise returned and Invoke the rest. Otherwise wait for promise to resolve before updating URL and invoking other listeners.

__Provided examples__

The provided examples are synchronous. Mostly because there are many ways to handle async data and I don't want to influence implementation. I have started working on a React application with router5 and I am hoping some patterns will emerge.

I am not familiar enough with backbone / marionette. Is it easy enough to trigger a view update on a specific node? An would the suggestion on a delayed URL change be useful?

**Comment by fabien** [comment]
Thanks again for your explanation. However, I'm not sure why you would need to limit your listeners to a single callback. I'm pretty sure you can step to each one, and check `can`.

There are two candidates for routing in my next projects, one is Router5, the other is https://github.com/visionmedia/page.js - I currently gravitate towards the flexibility the 'middleware' approach PageJS gives me, but the hierarchical nature of Router5 is also very appealing.

It's not fully clear to me how Router5 will fit my workflow, but the flow you outline above does sound compelling.

**Comment by troch** [comment]
I will only limit the "node listener type" to one per node, which makes sense because you only have one instance (or none) of a node at any time: [http://router5.github.io/docs/listeners.html](http://router5.github.io/docs/listeners.html). Other two types of listeners (listening to a specific route change, and listening to any route change).

Well honoured to be up there with page.js :grinning:! As you said, the two approaches are very different.

**Comment by troch** [comment]
v0.2.0 just released, transition works now as follow:

- `canDeactivate` called from bottom to top, each function can return a synchronous result in the form of a boolean, or a thenable (promise), or invoke a callback (as last argument)
- `canActivate` called from top to bottom, same than canDeactivate for sync / async results
- Associated node listener called
    - If it returns true, a promise or has a callback in its function signature it will wait for completion for updating URL and invoking other listeners. If it fails (returns false, rejected promise or callback with error), then transition fails
    - If it returns nothing and no 3rd argument (callback), it will update URL synchronously and will call other listeners immediately.

For __thenables__, you can use native ES6 promises, a polyfill or a promise implementation like q.

**Comment by fabien** [comment]
This looks great, thanks! It should fit my workflow much better now. One page.js nicety that shouldn't be too hard to integrate would be to allow multiple node listeners like this:

`myRouter.addNode('admin', '/admin', canAccessAdmin, loadUser, loadContent)`

For now I'll probably add a helper myself, but perhaps you can consider this as well:

`myRouter.addNode('admin', '/admin', wrapListeners(canAccessAdmin, loadUser, loadContent))`

**Comment by troch** [comment]
There is a confusion about what a node listener is: it is not a route listener. A node listener is triggered when changes are happening below it. For example, if you go from `home` to `admin`, the registered node listener for the unnamed root node will be called. If you navigate to `admin.users` from `admin.log`, the node listener for admin will be called. That's a different approach than page.js.

Also this router is designed to be  just a router. It won't care about data loading: it only wants to know if it can navigate away from some routes, if it can go to some routes, and optionally it can receive feedback from node listeners to know if loading went fine.

It is from node listeners you can then load data for the desired route and trigger a view update.. You can do it in series, in parallel, with promises, etc... With the help of your favourite control flow library.

I wouldn't mix data loading with canActivate because if you navigate deep and a segment below prevents activation, you could have an expensive data loading for nothing.

In short, I won't implement such a thing because it is outside the scope of this router :-).

**Comment by fabien** [comment]
Thanks, all clear! I'll see how it goes ...

**Comment by fabien** [comment]
@troch I'm trying to get a barebones setup working, and it looked like I succeeded. However, with `useHash` enabled, and with the router started, I don't seem to get notifications (of `addListener(fn)`) when reloading the page. The examples on your website do work like expected, so maybe Im missing something?

**Comment by fabien** [comment]
@troch could you please review the examples at http://router5.github.io/docs/listeners.html? It appears that there are some confusing errors there. I'm currently scratching my head over this:

var router = new Router5()
    .setOption('useHash', true)
    .setOption('defaultRoute', 'index')
    .addNode('index',      '/main')
    // Users
    .addNode('users',      '/users')
    .addNode('users.view', '/view/:id')
    .addNode('users.list', '/list')
    // Orders
    .addNode('orders',           '/orders')
    .addNode('orders.completed', '/completed')
    .addNode('orders.pending',   '/pending')
    .addNode('orders.view',      '/details/:id')
    
// Listeners

router.addListener(function(toState, fromState) {
    console.log('Root:', toState.name, fromState && fromState.name);
});

router.addNodeListener('users', function(toState, fromState) {
    console.log('Node:', toState.name, fromState && fromState.name);
});

router.addRouteListener('users.list', function(toState, fromState) {
    console.log('Route:', toState.name, fromState && fromState.name);
});

```

When navigating from `index` to `users.list`, shouldn't the `addNodeListener('users', fn)` be called, because it is passing through this node?

**Comment by fabien** [comment]
With the example above, when I add the following:

```
router.addNodeListener('users.list', function(toState, fromState) {
    console.log('Node:', toState.name, fromState && fromState.name);
    return false; // end here
});
```

It works only for the first (direct) hit on `users.list` - when I navigate to `index` and then back to `users.list` it is bypassed, and the route listener is executed regardless.

**Comment by troch** [comment]
- No node Listeners will be invoked on the first route loading (because it doesn't come from a previous route).
- When going from `index` to `users.list`, `addNodeListener('', fn)` (`index` is deactivated, `users` + `users.list` are activated: the intersection node between the two is the unamed root node.
- It normally shouldn't redirect to the default route if a route is matching. Possible bug?
- `router.addNodeListener('users.list)'`will never be called unless you go from `users.list.A` to `users.list.B`. If you return false in a node listener, the router will consider you are wanting to fail the transition.

**Comment by troch** [comment]
I have a bug to fix on start when there is a matching route => it should call `canActivate` on activating segments. It currently doesn't.

**Comment by fabien** [comment]
@troch did you try my example above - there's definitely some wonkyness there. I'm sorry, but I think Router5 sounds great in theory, but in practice it just doesn't work for me.

1. I would assume you'd envision coming from the defaultRoute, or just /
2.  `addNodeListener('', fn)`, I don't get why this is not `addNodeListener('users', fn)`
3. I think it's a bug, `/` is a catch-all it seems, it goes away when my index is `/main` for example
4. I can affirm that it *is* called, even without any sub-routes.

**Comment by fabien** [comment]
@troch thanks again, I really *do* want to use Router5, I'll wait until you fix the bug :). Main main gripe would be that  *No node Listeners will be invoked on the first route loading* eg. on start. 

Is this related to this bug?

**Comment by troch** [comment]
- Did you forget to `start()` in your example?
- The bug doesn't concern node listeners, it concern route activation on start. I'll work on fixing the bug later tonight, once everyone sleeps :zzz:
- I am positive about node listeners and calling `addNodeListener('', fn)`. I am not sure how else I can explain. Router5 is good for components tree, and with node listeners you can target from where you need to re-render your tree. So if you go from `index` to `users.list`, `users` doesn't exist yet so you can't re-render from `users`. Instead you have to re-render from above `index`: the unnamed root node. On `start`, you just need to grab `getState()`, possibly prepare some data and render your app fully. Node listeners are just here for re-renders.

**Comment by fabien** [comment]
- Yes, I omitted `start()` when I pasted the code here.
- OK, thanks
- I'm afraid that you have a very different idea of how the flow within the route tree happens. While Router5 appears to come really close to what I want, it is fundamentally different than I expected.

Basically, in my view, the flow would be somewhat like this when navigating from `index to `users.list`:

- index is a subnode, so go back to root first (perhaps check for a `''` node listener here)
- from root, pass through `users` - if and when the return value of (the optional) node listener here allows, proceed
- from `users` to `users.list` - there might be a node listener setup on `users.list`, verify it (promise, falsy)
- finally, reach `users.list` - if there's a route listener here, execute it

**Comment by troch** [comment]
> It works (blocks) only for the first (direct) hit on users.list - when I navigate (using links on the page) to index and then back to users.list it is bypassed, and the route listener is executed regardless.

Don't use raw hyperlinks on the page. You can't update the URL directly, it will confuse the router. You need to add a custom behaviour to your links (here is the example with React: https://github.com/router5/router5-react/blob/master/modules/link-factory.js).

```javascript
function onClick() {
    evt.preventDefault()
    router.navigate(routeName, routeParams)
}
```

For building your hyperlink (setting its href attribute):

```
var href =  router.buildUrl(routeName, routeParams)
```

For node listeners, they are __listeners__ <--- that's the really important word here, not actions triggered on a route change. From the beginning you are expecting router5 to proceed like page.js and execute a list of middle-ware functions in series. There is only one node listener called per route change, only one, no more than one. From there you organise your control flow for data, roll your view down the tree.

I am being patient and dedicating time to help, please don't be condescend regarding what you are "expecting".

**Comment by fabien** [comment]
I already used `onClick` handlers and `router.navigate`. It doesn't matter, there's no change... no matter what I do, returning `false` (or not calling the callback) in the `users` node listener is bypassed when navigating from `index` to `users.list`. Which seems to go completely against the concept of the flow within a tree. I'm sorry, but a tree has a definite serial flow to it IMHO...

I just wanted to explain things from my perspective - without any page.js preconceptions, nor expectations from you. 

Even when I read your emphasized explanation above, I cannot help but wonder why (or even if) we actually differ to this extent. It is pretty much what I would describe, too.

I never implied to expect node listeners to be 'actions' - it's very clear to me that listeners are event-based 'handlers' (even that term is tricky, in our discussion I guess), not endpoints/actions or something like that. 

We just seem to expect a different flow - instead of the logical path trough the tree, passing through nodes, you have something different in mind.

Unfortunately, it doesn't work for me after all. Thanks for your help, though & good luck with Router5.

**Comment by troch** [comment]
After reading this again, I think your misunderstanding comes from misinterpreting what the tree is: the real tree is your view (components + DOM). The router is just a representation of that tree. Your view is the one having a _serial flow_. The router is just there to tell you where to start your _serial flow_ from and have the cheapest re-rendering, but is not designed to orchestrate it.
